go/tasks/plugins/k8s/sagemaker/plugin.go

```diff
@@ -0,0 +1,163 @@
+package sagemaker
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/lyft/flyteplugins/go/tasks/pluginmachinery/utils"
+
+	flyteIdlCore "github.com/lyft/flyteidl/gen/pb-go/flyteidl/core"
+	"github.com/lyft/flytestdlib/logger"
+
+	"github.com/pkg/errors"
+	"k8s.io/client-go/kubernetes/scheme"
+
+	"github.com/lyft/flyteplugins/go/tasks/pluginmachinery"
+
+	commonv1 "github.com/aws/amazon-sagemaker-operator-for-k8s/api/v1/common"
+	hpojobv1 "github.com/aws/amazon-sagemaker-operator-for-k8s/api/v1/hyperparametertuningjob"
+	trainingjobv1 "github.com/aws/amazon-sagemaker-operator-for-k8s/api/v1/trainingjob"
+	pluginsCore "github.com/lyft/flyteplugins/go/tasks/pluginmachinery/core"
+	"github.com/lyft/flyteplugins/go/tasks/pluginmachinery/k8s"
+)
+
+// Sanity test that the plugin implements method of k8s.Plugin
+var _ k8s.Plugin = awsSagemakerPlugin{}
+
+type awsSagemakerPlugin struct {
+	TaskType pluginsCore.TaskType
+}
+
+func (m awsSagemakerPlugin) BuildIdentityResource(_ context.Context, _ pluginsCore.TaskExecutionMetadata) (k8s.Resource, error) {
+	if m.TaskType == trainingJobTaskType || m.TaskType == customTrainingJobTaskType {
+		return &trainingjobv1.TrainingJob{}, nil
+	}
+	if m.TaskType == hyperparameterTuningJobTaskType {
+		return &hpojobv1.HyperparameterTuningJob{}, nil
+	}
+	return nil, errors.Errorf("The sagemaker plugin is unable to build identity resource for an unknown task type [%v]", m.TaskType)
+}
+
+func (m awsSagemakerPlugin) BuildResource(ctx context.Context, taskCtx pluginsCore.TaskExecutionContext) (k8s.Resource, error) {
+
+	// Unmarshal the custom field of the task template back into the HyperparameterTuningJob struct generated in flyteidl
+	if m.TaskType == trainingJobTaskType {
+		return m.buildResourceForTrainingJob(ctx, taskCtx)
+	}
+	if m.TaskType == customTrainingJobTaskType {
+		return m.buildResourceForCustomTrainingJob(ctx, taskCtx)
+	}
+	if m.TaskType == hyperparameterTuningJobTaskType {
+		return m.buildResourceForHyperparameterTuningJob(ctx, taskCtx)
+	}
+	return nil, errors.Errorf("The SageMaker plugin is unable to build resource for unknown task type [%s]", m.TaskType)
+}
+
+func (m awsSagemakerPlugin) getEventInfoForJob(ctx context.Context, job k8s.Resource) (*pluginsCore.TaskInfo, error) {
+
+	var jobRegion, jobName, jobTypeInURL, sagemakerLinkName string
+	if m.TaskType == trainingJobTaskType {
```

Comment:
should we move these parts as submethods in the specific files? co-located code?

Reply:
I don't quite get what you mean. Mind to elaborate?

Reply:
So when task type == custom that code block could be in the custom.go file

Reply:
oh I see

Reply:
Done refactoring